design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+           ksqlDB environments. This allows users to version control their ksqlDB schema; recreate
+           their schema from scratch; and migrate their current schema to newer versions._
+
+## Motivation and background
+
+Schema migrations (also database migrations) refers to the process of performing updates or rollbacks on a database schema to a
+newer or older schema version. This practice is pretty common in any application lifecycle. Users use a version control for their
+applications which allow them to know if an application requires an upgrade or rolling back a buggy change. The same is necessary
+with a database schema. Users look for ways to version control schema changes along the application lifecycle.
+
+Existing tools exist to perform schema migrations on any database (MySQL, Postgres, Oracle, etc). These tools allow users to
+version the database schema, then automate the process to easily upgrade the schema to a newer version. If a schema is buggy,
+then users can rollback to the previous schema version. One more reason for using these tools is the integration with any
+CI/CD system, which allows users to test and verify a schema upgrade will work as expected.
+
+You can learn more about these existing tools:
+- [Flyway](https://flywaydb.org/)
+- [Liquibase](https://www.liquibase.org/)
+- [DBGeni](http://dbgeni.appsintheopen.com/index.html)
+
+ksqlDB requires the same kind of integration with CI/CD systems and automation to deploy ksqlDB schema changes in any environment.
+Some tools, like `Flyway`, support plugins to integrate schema migrations with a different non-supported database. However, our
+syntax is exclusively to ksqlDB; tools require JDBC support; and tools have several features that we cannot support
+(i.e. transactions & databases); so this integration becomes complex or unsupported. For that reason, ksqlDB must have its own
+schema migrations tool to provide the same migrations benefits to users who want to easily deploy and automate ksqlDB schema
+changes in all their ksqlDB environments.
+
+These benefits include:
+- Version the ksqlDB schema environments
+- Integrate ksqlDB schema changes with CI/CD environments
+- Simplify schema evolution changes across different environments
+
+This KLIP proposes a new tool for schema migrations. You will learn about the design aspects of the tool, and the features
+to support for a basic schema migration process.
+
+## What is in scope
+
+* Discuss design details for a new tool that provides schema migrations support for ksqlDB
+
+    Basic features to support:
+    - New CLI and API that can easily integrate with CI/CD environments
+    - Apply migrations on any ksqlDB environments
+    - Rollback migrated schemas to previous versions
+    - Version control ksqlDB schema
+
+* Provide a test plan that validates the new tool will meet the product requirements
+
+
+## What is not in scope
```

Comment:
Is squashing (periodically combining all previous migrations into one) out of scope? I don't think it needs to be a v1 thing, but this is historically important because over time, the value of having all migrations around diminishes.

Reply:
It is out of scope for this tool. The proposal for squashing is to use a different tool that will export the ksqlDB metastore to a SQL file. Squashing will be a combination of exporting the metastore and replace the migrations files with this new file.

If we want to integrate squashing into the migrations tool, then we'll have to support the testing framework (YATT) in order to run every migration in a testing environment, then export the current env as SQL file. We still need the metastore export tool.

Reply:
Makes sense, SGTM.